libs/libmysqlxx/CMakeLists.txt

```diff
@@ -26,25 +26,21 @@ add_library (mysqlxx
 
 add_dependencies (mysqlxx common)
 
-set (Z_HINTS "/usr/local/opt/zlib/lib")
-if (USE_STATIC_LIBRARIES)
-	find_library (Z_LIB libz.a HINTS ${Z_HINTS})
-else ()
-	find_library (Z_LIB z HINTS ${Z_HINTS})
-endif ()
-
 set(PLATFORM_LIBS ${CMAKE_DL_LIBS})
 
 if (USE_STATIC_LIBRARIES)
 	set (MYSQLCLIENT_LIB ${CMAKE_CURRENT_BINARY_DIR}/libmysqlclient.a)
-	target_link_libraries (mysqlxx common ${MYSQLCLIENT_LIB} ${OPENSSL_LIBRARIES} ${Z_LIB} ${PLATFORM_LIBS})
 	add_custom_command (
 		OUTPUT ${MYSQLCLIENT_LIB}
 		COMMAND ${CMAKE_CURRENT_SOURCE_DIR}/patch.sh ${STATIC_MYSQLCLIENT_LIB} ${MYSQLCLIENT_LIB}
 		WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
 		COMMENT "Patching mysqlclient library.")
 	add_custom_target (our_mysql_client DEPENDS ${MYSQLCLIENT_LIB})
 	add_dependencies (mysqlxx our_mysql_client)
+	set (Z_LIB zlibstatic)
+else ()
+	set (Z_HINTS "/usr/local/opt/zlib/lib")
+	find_library (Z_LIB z HINTS ${Z_HINTS})
 endif ()
 
 target_link_libraries (mysqlxx common ${MYSQLCLIENT_LIB} ${OPENSSL_LIBRARIES} ${Z_LIB} ${PLATFORM_LIBS})
```

Comment:
The reason I inserted if (USE_STATIC_LIBRARIES) everywhere is that libmysqlclient.so is dynamically linked to libz.so. As everything is linked together in the clickhouse executable, it means that we need to link with libz.so everywhere else.

What is the policy here? To what extent does the build system support dynamic linking? Why is it needed?